google-cloud-logging/src/main/java/com/google/cloud/logging/BatchingWriter.java

```diff
@@ -0,0 +1,152 @@
+ * limitations under the License.
+ */
+package com.google.cloud.logging;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.common.base.Preconditions;
+import com.google.logging.v2.WriteLogEntriesRequest;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+class BatchingWriter {
+
+  interface Rpc {
+    ApiFuture<Void> call(WriteLogEntriesRequest request);
+  }
+
+  private final Rpc rpc;
+  private final int batchSize;
+  private final Semaphore pending;
+  private final WriteLogEntriesRequest requestPrototype;
+  private final ScheduledThreadPoolExecutor executor;
+
+  // Use Boolean, since concurrent maps don't support nulls.
+  private final ConcurrentHashMap<ApiFuture<Void>, Boolean> pendingWrites =
+      new ConcurrentHashMap<>();
+
+  private final ArrayList<com.google.logging.v2.LogEntry> currentBatch;
+
+  private Future<?> flushJob = null;
+
+  BatchingWriter(
+      Rpc rpc,
+      int batchSize,
+      int maxPending,
+      WriteLogEntriesRequest requestPrototype,
+      ScheduledThreadPoolExecutor executor) {
+    Preconditions.checkArgument(batchSize > 0, "batchSize must be positive");
+    this.batchSize = batchSize;
+    Preconditions.checkArgument(maxPending > 0, "maxPending must be positive");
+    this.pending = new Semaphore(maxPending);
+
+    this.rpc = Preconditions.checkNotNull(rpc);
+    this.requestPrototype = Preconditions.checkNotNull(requestPrototype);
+    this.executor = Preconditions.checkNotNull(executor);
+    this.currentBatch = new ArrayList<>(batchSize);
+  }
+
+  synchronized void startJob() {
+    Preconditions.checkArgument(flushJob == null, "job already started");
+    flushJob =
+        executor.scheduleWithFixedDelay(
+            new Runnable() {
+              @Override
+              public void run() {
+                initFlush();
+              }
+            },
+            100,
+            100,
+            TimeUnit.MILLISECONDS);
+  }
+
+  synchronized void stopJob() {
+    Preconditions.checkArgument(flushJob != null, "job not started");
+    flushJob.cancel(false);
+  }
+
+  void add(com.google.logging.v2.LogEntry entry) throws InterruptedException {
+    pending.acquire(1);
+    synchronized (currentBatch) {
+      currentBatch.add(entry);
+      if (currentBatch.size() == batchSize) {
+        final WriteLogEntriesRequest request =
+            requestPrototype.toBuilder().addAllEntries(currentBatch).build();
+        currentBatch.clear();
+
+        // Whoever calls send serializes the proto; so we do it off-thread.
+        // This gives better CPU utilization if there are few producer threads
+        // on a many-core machine.
+        executor.execute(
```

Comment:
Actually, a little bug here. If we add, then initFlush, initFlush might happen before send. A fix: create a "surrogate" future and put it in the map before unlocking. I'll perf test this.